### PR TITLE
#184347026 update user roles

### PR DESCRIPTION
**Pull request by Aline-Uwera**
#### What does this PR do?
Admin Should be able to set roles/permissions
#### Description of Task to be completed?
Admin should be able to assign various users to various roles. 
#### How should this be manually tested?
1. Checkout to this branch: `feature/user-login-184347020`
2. Run the command `npm install` 
3. Open browser and visit http://localhost:5000/api-docs/#/
4.  Test functionalities of user/ updateRole route
#### What are the relevant pivotal tracker stories?
[184347026](https://www.pivotaltracker.com/story/show/184347026)
#### Screenshoots
N/A
#### Questions
N/A

### CODE REVIEW

**Review by Fidela1 — APPROVED** [review]

File: package.json [hunk 1]
```diff
@@ -16,7 +16,7 @@
   },
   "betterScripts": {
     "dev": {
-      "command": "nodemon --exec npx babel-node server",
+      "command": "npm run seeding && nodemon --exec npx babel-node server",
```
Comment (ncutixavier) [inline package.json]: You don't need seeding when running in server

File: src/controllers/user.controller.js [hunk 2]
```diff
@@ -36,4 +37,31 @@ export class UserController {
...
+		} catch (error) {
+			next(error);
+		}
+	}
+
+	static async updateRole(req, res) {
+		try {
+			if(!req.body.role){
+				return res.status(400).json({ message: 'Missing required field' });
+			}
```
Comment (ncutixavier) [inline src/controllers/user.controller.js]: This should be implemented in user validation

File: src/controllers/user.controller.js [hunk 3]
```diff
@@ -36,4 +37,31 @@ export class UserController {
...
+			let data = {id:req.params.id,role:req.body.role};
```
Comment (ncutixavier) [inline src/controllers/user.controller.js]: Let's enforce destructuring assignment

File: src/controllers/user.controller.js [hunk 4]
```diff
@@ -36,4 +37,31 @@ export class UserController {
...
+
+	static async updateRole(req, res) {
+		try {
+			if(!req.body.role){
+				return res.status(400).json({ message: 'Missing required field' });
+			}
+			let data = {id:req.params.id,role:req.body.role}; 
+			const response = await UpdateRoleServices.updateRole(data);
+			if (response == false) {
+				return res.status(404).json({ error: `User with ID = ${req.params.id} does not exist` });
```
Comment (ncutixavier) [inline src/controllers/user.controller.js]: You should use already implemented middleware to check if the user exists. Ask Yves

File: src/controllers/user.controller.js [hunk 5]
```diff
@@ -36,4 +37,31 @@ export class UserController {
...
+			} else {
+				return res.status(200).json({ message: 'Updated successfully' });
+			}
+		} catch (error) {
+			return res.status(500).json({ error: error });
+		}
```
Comment (ncutixavier) [inline src/controllers/user.controller.js]: Let's follow the same convention used on register in the catch block. Ask Abdul

File: src/middlewares/isAdmin.js [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+                    status: 401,
+                    message: 'Only admin is allowed to perform this task',
+                    });
+            }else{
+                next();
+            }
+        });
+    
+    };
+
```
Comment (ncutixavier) [inline src/middlewares/isAdmin.js]: Let's have global middleware which will support protecting route to be accessed by a specified role

`You can implement this in auth.middle...`
image

`Use this middleware in route after checking if the user is authenticated
image

File: src/routes/api/user.routes.js [hunk 7]
```diff
@@ -13,4 +15,16 @@ router.post(
 
 router.post('/login', getExistUser, checkUserVerified, checkValidPassword,UserController.loginUser);
 
+router.get(
+  '/', 
+  isAdmin,
```
Comment (ncutixavier) [inline src/routes/api/user.routes.js]: I am expecting two middlewares here to replace isAdmin
- protectRoute
- restrictedTo(<roles>)

File: src/services/updateRoleServices.js [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+            where: {
+              id: user.id,
+              role: user.role,
+            },
+          }
+        );
+        return true;
+    }
+  }
+}
```
Comment (ncutixavier) [inline src/services/updateRoleServices.js]: - Service should only hold the query. Ask Yves for help
- This is user related service

**Review by ncutixavier — CHANGES_REQUESTED** [review]
All check should be successful passed

File: src/routes/api/user.routes.js (hunk 7 above)
Comment (Aline-Uwera) [reply src/routes/api/user.routes.js]: For protectRoute, we have auth.middleware , I think they may be performing same task

**Review by Abdden — APPROVED** [review]

**Review by ndatumuremyi — APPROVED** [review]

File: src/controllers/user.controller.js [hunk 9]
```diff
@@ -42,10 +44,39 @@ export class UserController {
...
+	}
+
+	static async updateRole(req, res) {
+		try {
+			let data = {id:req.params.id};
+			let updatedRole = req.body.role;
+			const user = await User.findOne({ where: { id: data.id } });
+			if (!user) {
+				return res.status(404).json({ error: `User with ID = ${req.params.id} does not exist` });
```
Comment (ncutixavier) [inline src/controllers/user.controller.js]: Use already implemented middleware

File: src/controllers/user.controller.js [hunk 10]
```diff
@@ -42,10 +44,39 @@ export class UserController {
...
+		}
+	}
+
+	static async getAllUsers (req, res, next) {
+		try {
+			const users = await models.User.findAll();
+			res.status(200).json({status: 200, message: 'All Users retrieved successfully', data: users});
+		} catch (error) {
+			next(error);
+		}
```
Comment (ncutixavier) [inline src/controllers/user.controller.js]: - Next should not used in the controller
- Follow the same convention implemented on Login. Ask Yves

File: src/middlewares/restrictedTo.js [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+export const restrictTo = (...roles) => {
+  return (req, res, next) => {
+    if (!roles.includes(req.user.role)) {
+      return res.status(403).json({
+        message: 'You are not allowed to perform this task',
+      });
+    }
+    next();
+  };
+};
```
Comment (ncutixavier) [inline src/middlewares/restrictedTo.js]: Let's follow the file naming convention

File: src/services/user.service.js [hunk 12]
```diff
@@ -1,6 +1,21 @@
...
+    static async updateRole(user,updatedRole) {
+        return await User.update(
+            { role: updatedRole },
+            {
+                where: {
+                  id: user.id,
+                  role: user.role,
+                },
+            }
+        );
```
Comment (ncutixavier) [inline src/services/user.service.js]: This should be a reusable service to update users, not to be specific for role.

File: src/validations/user/role.validation.js [hunk 13]
```diff
@@ -0,0 +1,16 @@
+import Joi from 'joi';
+
+const validateRole = async (req, res, next) => {
+  const roleSchema= Joi.object({
+    role: Joi.string()
+      .valid('admin', 'seller', 'buyer')
```
Comment (ncutixavier) [inline src/validations/user/role.validation.js:6]: Are you sure these are the only roles you need?

**Review by irakozetony — APPROVED** [review]
Looks good to me!

**Review by ncutixavier — CHANGES_REQUESTED** [review]

File: src/middlewares/user.middleware.js [hunk 14]
```diff
@@ -28,6 +28,17 @@ export const checkUserVerified = async (req, res, next) => {
   }
   next();
 };
+export const checkIfUserExistById = async(req, res, next) => {
+
+    const id = req.params.id;
```
Comment (ncutixavier) [inline src/middlewares/user.middleware.js:33]: The ID should be passed in the body not the params

**Review by ncutixavier — CHANGES_REQUESTED** [review]

**Review by ncutixavier — APPROVED** [review]

**Pull request merged**